libs/core/algorithms/include/hpx/parallel/algorithms/uninitialized_relocate.hpp

```diff
@@ -644,5 +830,87 @@ namespace hpx::experimental {
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator_v<BiIter1> &&
+                hpx::traits::is_iterator_v<BiIter2>
+            )>
+        // clang-format on
+        friend BiIter2 tag_fallback_invoke(uninitialized_relocate_backward_t,
+            BiIter1 first, BiIter1 last,
+            BiIter2 dest_last) noexcept(util::detail::relocation_traits<BiIter1,
+            BiIter2>::is_noexcept_relocatable_v)
+        {
+            static_assert(hpx::traits::is_bidirectional_iterator_v<BiIter1> &&
+                "The 'first' and 'last' arguments must meet the requirements "
+                "of bidirectional iterators.");
+            static_assert(hpx::traits::is_bidirectional_iterator_v<BiIter2>,
+                "The 'dest_last' argument must meet the requirements of a "
+                "bidirectional iterator.");
+            static_assert(util::detail::relocation_traits<BiIter1,
+                              BiIter2>::valid_relocation,
+                "Relocating from this source type to this destination type is "
+                "ill-formed");
+            // if count is representing a negative value, we do nothing
+            if (first == last)
+            {
+                return dest_last;
+            }
+
+            return parallel::util::get_second_element(
+                hpx::parallel::detail::uninitialized_relocate_backward<
+                    parallel::util::in_out_result<BiIter1, BiIter2>>()
+                    .call(hpx::execution::seq, first, last, dest_last));
+        }
+
+        // clang-format off
+        template <typename ExPolicy, typename BiIter1, typename BiIter2,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy_v<ExPolicy> &&
+                hpx::traits::is_iterator_v<BiIter1> &&
+                hpx::traits::is_iterator_v<BiIter2>
+            )>
+        // clang-format on
+        friend typename hpx::parallel::util::detail::algorithm_result<ExPolicy,
+            BiIter2>::type
+        tag_fallback_invoke(uninitialized_relocate_backward_t,
+            ExPolicy&& policy, BiIter1 first, BiIter1 last,
+            BiIter2 dest_last) noexcept(util::detail::relocation_traits<BiIter1,
+            BiIter2>::is_noexcept_relocatable_v)
+        {
+            static_assert(hpx::traits::is_bidirectional_iterator_v<BiIter1>,
+                "The 'first' and 'last' arguments must meet the requirements "
+                "of bidirectional iterators.");
+            static_assert(hpx::traits::is_bidirectional_iterator_v<BiIter2>,
+                "The 'dest' argument must meet the requirements of a "
+                "bidirectional iterator.");
+            static_assert(util::detail::relocation_traits<BiIter1,
+                              BiIter2>::valid_relocation,
+                "Relocating from this source type to this destination type is "
+                "ill-formed");
+
+            auto count = std::distance(first, last);
```

Comment:
@hkaiser @Pansysk75 @gonidelis Considering https://github.com/STEllAR-GROUP/hpx/pull/6364#issuecomment-1771644408, it may be sensible to include a warning here such as:
"Calling `uninitialized_relocate_backward` with a non-sequential execution policy does not guarantee the order of execution of the relocations to be backward. Use on overlapping ranges could yield undefined behavior."

However, I am not sure how we would issue such a warning.

Reply:
There is no need to do that. The same warning applies to `std::copy` and `std::move`. There, it is up to the user to ensure these preconditions.

Reply:
@Quuxplusone Thank you for the detailed write-up, and sorry for my late response. 

I think my previous comment:  https://github.com/STEllAR-GROUP/hpx/pull/6364#issuecomment-1771644408, is incorrect and there does exist a way to parallelize the relocation of overlapping ranges, maybe like in the following image: (example for 2 threads and an overlapping offset of 1) 

![image](https://github.com/STEllAR-GROUP/hpx/assets/63551950/e6bac3cf-d30c-409a-a706-097d80709885)

Supposing this is true, I see no reason to avoid offering this as a feature. Furthermore, I don't see why we would need non-overlapping ranges to be a precondition for relocation!

> I think you could fix the latter inconsistency by having hpx::parallel::detail::parallel_uninitialized_relocate_n check for overlap

I think for the time being we will do this, only parallelizing the guaranteed safe case (contiguous iterator + no overlap), and have the rest be sequenced, but definitely add proper parallelization support on the todo list. @hkaiser What do you think?

> maybe you'd rather that P1144's std::uninitialized_relocate_backward should forbid overlap .... If so, I'd like to know what's your plan for implementing things like {parallel_,}vector::erase

For the time being we do not implement parallel data structures, but there is a draft PR for using relocation inside `hpx::detail::small_vector` (which is like `static_vector` except it grows to dynamic storage after running out of static space)  (https://github.com/isidorostsa/hpx_fork/pull/9).